File: leetcode/70/solution.go
// 70. Climbing Stairs
// https://leetcode.com/problems/climbing-stairs/
package lc70

func climbStairs(n int) int {
	// Initialize slice dp with 0 (we need it with the length of n+1 )
	dp := make([]int, n+1)
	i := 0
	for i < n {
		dp[i] = 0
		i++
	}

	var dfs func(x int) int
	dfs = func(x int) int {
		// If x is 1 or 2, just return it.
		if x <= 2 {
			return x
		}
		// If we have already populate x, then return it.
		if 0 < dp[x] {
			return dp[x]
		}
		// Now x is greater than 2 and we don't have the result of dp[x].
		// Populate result first and then return it.
		dp[x] = dfs(x-2) + dfs(x-1)
		return dp[x]

	}

	return dfs(n)
}

File: leetcode/70/solution.py
"""
70. Climbing Stairs
https://leetcode.com/problems/climbing-stairs/
"""
from functools import lru_cache


class Solution:
    def climbStairs(self, n: int) -> int:
        @lru_cache()
        def dfs(x: int) -> int:
            if x == 1:  # we have only 1 pattern (that is, [1])
                return 1
            if x == 2:  # we have only 2 patterns ([1,1] and [2])
                return 2
            # otherwise, search and combine the result of n - 2 and n - 1
            return dfs(x - 2) + dfs(x - 1)

        return dfs(n)
